File: promotion/coupon-service/src/main/java/org/study/couponservice/config/UserIdInterceptor.java
package org.study.couponservice.config;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.hibernate.annotations.Comment;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

@Slf4j
@Component
public class UserIdInterceptor implements HandlerInterceptor {
    private static final String USER_ID_HEADER = "X-USER-ID";
    private static final ThreadLocal<Long> currentUserId = new ThreadLocal<>();

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
        String userIdStr = request.getHeader(USER_ID_HEADER);
        log.info("Received X-USER-ID: {}", userIdStr);

        if (userIdStr == null || userIdStr.isEmpty()) {
            throw new IllegalStateException("요청 헤더에 X-USER-ID가 누락되었습니다");
        }

        try {
            currentUserId.set(Long.parseLong(userIdStr));
            return true;
        } catch (NumberFormatException e) {
            throw new IllegalStateException("요청 헤더 X-USER-ID의 값이 올바른 숫자 형식이 아닙니다.");
        }
    }

    @Override
    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex) throws Exception {
        currentUserId.remove();
    }

    public static Long getCurrentUserId() {
        Long userId = currentUserId.get();
        if (userId == null) {
            throw new IllegalStateException("User ID not found in current context");
        }
        return userId;
    }
}

File: promotion-kotlin/coupon-service-kotlin/src/main/kotlin/com/study/couponservice/config/UserIdInterceptor.kt
package com.study.couponservice.config

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor

@Component
class UserIdInterceptor : HandlerInterceptor {

    companion object {
        private const val USER_ID_HEADER = "X-USER-ID"
        private val currentUserId: ThreadLocal<Long> = ThreadLocal()

        fun getCurrentUserId(): Long {
            return currentUserId.get() ?: throw IllegalStateException("User ID not found in current context")
        }
    }

    override fun preHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any
    ): Boolean {
        val userIdStr = request.getHeader(USER_ID_HEADER)
            ?: throw IllegalStateException("X-USER-ID header is required")
        currentUserId.set(userIdStr.toLongOrNull() ?: throw IllegalStateException("Invalid X-USER-ID format"))
        return true
    }

    override fun afterCompletion(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        ex: Exception?
    ) {
        currentUserId.remove()
    }
}
